## Assignment 2 

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from sklearn.linear_model import LinearRegression
import statsmodels.api as sm

/anaconda3/lib/python3.6/site-packages/statsmodels/compat/pandas.py:56: FutureWarning: The pandas.core.datetools module is deprecated and will be removed in a future version. Please use the pandas.tseries module instead.
  from pandas.core import datetools


In [52]:
df = pd.read_csv("homes.csv")

In [53]:
df['age'] = 2001 - df['year']
df.head()

price  area  beds  baths  garage  year  style  lotsize   ac pool quality  \
0  360000  3032     4      4       2  1972      1    22221  YES   NO  MEDIUM   
1  340000  2058     4      2       2  1976      1    22912  YES   NO  MEDIUM   
2  250000  1780     4      3       2  1980      1    21345  YES   NO  MEDIUM   
3  205500  1638     4      2       2  1963      1    17342  YES   NO  MEDIUM   
4  275500  2196     4      3       2  1968      7    21786  YES   NO  MEDIUM   

  highway  age  
0      NO   29  
1      NO   25  
2      NO   21  
3      NO   38  
4      NO   33

### 1. Fit a model with one predictor only (the number of bedrooms). What is the estimated increase in the average price of a house for each additional bedroom?

In [4]:
df1 = pd.read_csv("homes.csv", usecols=['price','beds'])

In [5]:
price = df1.price
beds = df1.beds

In [6]:
beds = sm.add_constant(beds)
model = sm.OLS(price, beds).fit()
predictions = model.predict(beds)
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  price   R-squared:                       0.171
Model:                            OLS   Adj. R-squared:                  0.169
Method:                 Least Squares   F-statistic:                     107.1
Date:                Fri, 18 Oct 2019   Prob (F-statistic):           5.89e-23
Time:                        14:36:38   Log-Likelihood:                -6868.9
No. Observations:                 522   AIC:                         1.374e+04
Df Residuals:                     520   BIC:                         1.375e+04
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const       8.281e+04   1.96e+04      4.218      0.000    4.42e+04    1.21e+05
beds         5.62e+04   5429.563     10.351      0.000    4.55e+04    6.69e+04
==============================================================================
Omnibus:                      189.593   Durbin-Watson:                   0.557
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              565.334
Skew:                           1.767   Prob(JB):                    1.73e-123
Kurtosis:                       6.676   Cond. No.                         13.8
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [7]:
model.params

const    82808.799613
beds     56200.083666
dtype: float64

#### Ans: The coefficient of 56200.08366 means that as the beds variable increases by 1, the predicted value of price increases by 56200.08366.

### 2. Fit the full model (with variables x1, . . . , x6).

#### a) Report the estimated regression coefficients (include a 99% confidence interval for each one).

In [8]:
X = df[['lotsize','area','beds','baths','age','garage']]
y = df['price']

In [9]:
X = sm.add_constant(X)
model = sm.OLS(y, X).fit()
model.conf_int(0.01) 
# 99% confidence interval for each one

0             1
const   -55431.280841  42016.204508
lotsize      0.851721      2.258260
area       107.735683    143.741523
beds    -22815.268707  -3267.519722
baths    -4471.425313  20446.528968
age      -2336.734049  -1222.487627
garage    7497.919937  37562.838640

In [10]:
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  price   R-squared:                       0.751
Model:                            OLS   Adj. R-squared:                  0.748
Method:                 Least Squares   F-statistic:                     259.4
Date:                Fri, 18 Oct 2019   Prob (F-statistic):          4.21e-152
Time:                        14:36:43   Log-Likelihood:                -6554.5
No. Observations:                 522   AIC:                         1.312e+04
Df Residuals:                     515   BIC:                         1.315e+04
Df Model:                           6                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const      -6707.5382   1.88e+04     -0.356      0.722   -4.37e+04    3.03e+04
lotsize        1.5550      0.272      5.717      0.000       1.021       2.089
area         125.7386      6.963     18.057      0.000     112.059     139.419
beds       -1.304e+04   3780.397     -3.450      0.001   -2.05e+04   -5614.498
baths       7987.5518   4818.957      1.658      0.098   -1479.680    1.75e+04
age        -1779.6108    215.487     -8.259      0.000   -2202.953   -1356.268
garage      2.253e+04   5814.344      3.875      0.000    1.11e+04     3.4e+04
==============================================================================
Omnibus:                      124.114   Durbin-Watson:                   1.327
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              384.582
Skew:                           1.110   Prob(JB):                     3.08e-84
Kurtosis:                       6.571   Cond. No.                     1.71e+05
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 1.71e+05. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

#### b) Find a 95% confidence interval for the average price of a house with garage for two cars, area of 2650 square feet, built in 1990, 24500 square feet size, three bedrooms, and, three bathrooms.

In [11]:
plist1 = ['constant','lotsize','area','beds','baths','age','garage']
plist1_v = [1,24500, 2650, 3, 3, 2001-1990, 2]

In [12]:
x = pd.DataFrame([plist1_v],columns = plist1)
x

constant  lotsize  area  beds  baths  age  garage
0         1    24500  2650     3      3   11       2

In [13]:
pred = model.get_prediction(x)

In [14]:
pred.summary_frame(alpha = 0.05)

mean      mean_se  mean_ci_lower  mean_ci_upper  obs_ci_lower  \
0  374920.532647  6511.365875  362128.426983   387712.63831  238431.17994   

    obs_ci_upper  
0  511409.885353

#### Ans: 362128.426983 ~ 387712.63831

#### c) Predict the price of a house when all predictors are equal to their median values.

In [15]:
df.median()

price      229900.0
area         2061.0
beds            3.0
baths           3.0
garage          2.0
year         1966.0
style           2.0
lotsize     22200.0
age            35.0
dtype: float64

In [16]:
plist2 = ['constant','lotsize','area','beds','baths','age','garage']
plist2_v = [1,22200, 2061, 3, 3, 35, 2]

In [17]:
x = pd.DataFrame([plist2_v],columns = plist2)
x

constant  lotsize  area  beds  baths  age  garage
0         1    22200  2061     3      3   35       2

In [18]:
pred = model.get_prediction(x)
pred.summary_frame()

mean      mean_se  mean_ci_lower  mean_ci_upper   obs_ci_lower  \
0  254573.357919  4595.921615   245544.29765  263602.418187  118385.144904   

    obs_ci_upper  
0  390761.570933

#### Ans: 254573.357919

### 3. What are the best and worst predictors, in terms of correlation with the house price?

In [19]:
corr = df.corr()
corr.style.background_gradient().set_precision(2)

#### Ans: area is the best predictor and lotsize is the worst predictor with price

### 4. What are the best and worst predictors, in terms of R2?

In [20]:
list1 = ['lotsize','area','beds','baths','garage','age']
df.head()

price  area  beds  baths  garage  year  style  lotsize   ac pool quality  \
0  360000  3032     4      4       2  1972      1    22221  YES   NO  MEDIUM   
1  340000  2058     4      2       2  1976      1    22912  YES   NO  MEDIUM   
2  250000  1780     4      3       2  1980      1    21345  YES   NO  MEDIUM   
3  205500  1638     4      2       2  1963      1    17342  YES   NO  MEDIUM   
4  275500  2196     4      3       2  1968      7    21786  YES   NO  MEDIUM   

  highway  age  
0      NO   29  
1      NO   25  
2      NO   21  
3      NO   38  
4      NO   33

In [21]:
data = {}
df_r2 = pd.DataFrame(data)
df_r2['column'] = list1

In [22]:
r2 = []
for i in range(0,len(list1)):
    X = df[list1[i]]
    y = df['price']
    X = sm.add_constant(X)
    model_r = sm.OLS(y, X).fit()
    r2.append(model_r.rsquared)
df_r2['R_squared'] = r2
df_r2

column  R_squared
0  lotsize   0.050252
1     area   0.671531
2     beds   0.170837
3    baths   0.467426
4   garage   0.333837
5      age   0.308599

#### Ans: area is the best and lotsize is the worst.

### 5.  Fit the model with the best subset of predictors (in terms of AIC). Report the fitted equation.

In [23]:
from sklearn.metrics import mean_squared_error
# find the all AIC from all subset. 
list2 = ['lotsize','area','beds','baths','age','garage']
X = df[list2]
y = df['price']

In [24]:
def get_sse(x, y):
    model = LinearRegression().fit(x, y)
    SSE = mean_squared_error(y, model.predict(x))*(len(y)-2)
    R_squared = model.score(x, y)
    return SSE, R_squared

In [25]:
SSE_list, feature_list, num_features = [],[],[]

In [26]:
df[list2].shape

(522, 6)

In [27]:
n = 522
p = 6

In [28]:
import itertools

In [29]:
m = len(X.columns)
list(range(1, m+1))

[1, 2, 3, 4, 5, 6]

In [30]:
for k in range(1, m+1):
    for subset in itertools.combinations(X.columns, k):
        temp = get_sse(X[list(subset)], y)
        SSE_list.append(temp[0])
        feature_list.append(subset)
        num_features.append(len(subset))

In [31]:
df = pd.DataFrame(num_features, columns = ['n_features'])
df['feature'] = feature_list
SSE_list2 = np.round(SSE_list, 2)
df['SSE'] = SSE_list2
m = n

In [32]:
df['AIC'] = n*np.log(df['SSE']/n) + 2*df['n_features']
df[:5]

n_features     feature           SSE           AIC
0           1  (lotsize,)  9.376809e+12  12327.251558
1           1     (area,)  3.242953e+12  11773.015318
2           1     (beds,)  8.186279e+12  12256.374411
3           1    (baths,)  5.258074e+12  12025.287719
4           1      (age,)  6.826165e+12  12161.528810

In [33]:
best_AIC = df.groupby(['n_features'])['AIC'].min()
df1 = df[df['AIC'] == best_AIC[1]]

In [34]:
for i in range(2, 7):
    df1 = df1.append(df[df['AIC'] == best_AIC[i]])
df1

n_features                                    feature           SSE  \
1            1                                    (area,)  3.242953e+12   
13           2                                (area, age)  2.781711e+12   
23           3                       (lotsize, area, age)  2.587262e+12   
46           4               (lotsize, area, age, garage)  2.513275e+12   
58           5         (lotsize, area, beds, age, garage)  2.467612e+12   
62           6  (lotsize, area, beds, baths, age, garage)  2.454517e+12   

             AIC  
1   11773.015318  
13  11694.931097  
23  11659.103920  
46  11645.958769  
58  11638.387401  
62  11637.610063

In [35]:
row6 = df1[df1.AIC == df1.AIC.min()]
row6
# take this subset as model

n_features                                    feature           SSE  \
62           6  (lotsize, area, beds, baths, age, garage)  2.454517e+12   

             AIC  
62  11637.610063

In [36]:
X = sm.add_constant(X)
model = sm.OLS(y, X).fit()
model.params

const      -6707.538167
lotsize        1.554990
area         125.738603
beds      -13041.394214
baths       7987.551827
age        -1779.610838
garage     22530.379289
dtype: float64

#### Ans: y = -6707.54 + 1.555*lotsize + 125.74*area - 13041.394*beds + 7987.55*baths - 1779.61*age + 22530.379*garage

### 6. Fit a linear regression model and report the fitted equations. (Consider now predictors lotsize and quality only)

In [37]:
#dummy
df4 = pd.read_csv("homes.csv", usecols=['lotsize','quality','price'])
df4.head()

price  lotsize quality
0  360000    22221  MEDIUM
1  340000    22912  MEDIUM
2  250000    21345  MEDIUM
3  205500    17342  MEDIUM
4  275500    21786  MEDIUM

In [38]:
credit1 = pd.get_dummies(df4,columns = ['quality'],
                         drop_first = True)
credit1.head()

price  lotsize  quality_LOW  quality_MEDIUM
0  360000    22221            0               1
1  340000    22912            0               1
2  250000    21345            0               1
3  205500    17342            0               1
4  275500    21786            0               1

In [39]:
X = credit1[['lotsize','quality_LOW','quality_MEDIUM']]
y = credit1['price']

In [40]:
X = sm.add_constant(X)
model2 = sm.OLS(y, X).fit()
model2.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  price   R-squared:                       0.679
Model:                            OLS   Adj. R-squared:                  0.677
Method:                 Least Squares   F-statistic:                     365.5
Date:                Fri, 18 Oct 2019   Prob (F-statistic):          2.04e-127
Time:                        14:37:02   Log-Likelihood:                -6621.1
No. Observations:                 522   AIC:                         1.325e+04
Df Residuals:                     518   BIC:                         1.327e+04
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
==================================================================================
                     coef    std err          t      P>|t|      [0.025      0.975]
----------------------------------------------------------------------------------
const           4.996e+05   1.24e+04     40.377      0.000    4.75e+05    5.24e+05
lotsize            1.6421      0.296      5.551      0.000       1.061       2.223
quality_LOW    -3.617e+05   1.14e+04    -31.818      0.000   -3.84e+05   -3.39e+05
quality_MEDIUM -2.666e+05   1.06e+04    -25.210      0.000   -2.87e+05   -2.46e+05
==============================================================================
Omnibus:                      114.013   Durbin-Watson:                   0.924
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              285.522
Skew:                           1.097   Prob(JB):                     9.99e-63
Kurtosis:                       5.883   Cond. No.                     1.42e+05
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 1.42e+05. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

In [41]:
model2.params

const             499607.815886
lotsize                1.642053
quality_LOW      -361721.638336
quality_MEDIUM   -266554.185858
dtype: float64

#### Ans: y = 499607.815886 + 1.642053*lotsize - -361721.638336*quality_LOW - 266554.185858*quality_MEDIUM

### 7. Predict the price of a house with lotsize 40000 square feet and medium quality.

In [42]:
df4 = pd.read_csv("homes.csv", usecols=['price','lotsize','quality'])
df4.head()

price  lotsize quality
0  360000    22221  MEDIUM
1  340000    22912  MEDIUM
2  250000    21345  MEDIUM
3  205500    17342  MEDIUM
4  275500    21786  MEDIUM

In [43]:
credit1 = pd.get_dummies(df4,columns = ['quality'],
                         drop_first = True)
credit1.head()

price  lotsize  quality_LOW  quality_MEDIUM
0  360000    22221            0               1
1  340000    22912            0               1
2  250000    21345            0               1
3  205500    17342            0               1
4  275500    21786            0               1

In [44]:
X = credit1[['lotsize','quality_LOW','quality_MEDIUM']]
y = credit1['price']
X = sm.add_constant(X)
model = sm.OLS(y, X.astype(float)).fit()

In [45]:
list2 = ['constant','lotsize','quality_LOW','quality_MEDIUM']
list2_v = [1,40000, 0,1]

In [46]:
x = pd.DataFrame([list2_v],columns = list2)
x

constant  lotsize  quality_LOW  quality_MEDIUM
0         1    40000            0               1

In [47]:
pred = model.get_prediction(x)

In [48]:
pred.summary_frame()

mean      mean_se  mean_ci_lower  mean_ci_upper   obs_ci_lower  \
0  298735.732619  6434.186782  286095.423979   311376.04126  144292.907077   

    obs_ci_upper  
0  453178.558161

#### Ans: 298735.732619

### 8.  Fit a full model for houses having between two to four bedrooms. Find a 95% prediction interval for the price of a high quality house with a garage for two cars, area of 3150 square feet, built in 1996, 26250 square feet size, two bedrooms, three bathrooms style 1, with AC, a pool, not close to a highway.

In [54]:
df5 = df[(df['beds'] >= 2) & (df['beds'] <= 4)]
df5 = df5.reset_index(drop=True)

In [55]:
credit1 = pd.get_dummies(df5,columns = ['quality','style','highway','pool','ac'],
                         drop_first = True)
credit1.head()

price  area  beds  baths  garage  year  lotsize  age  quality_LOW  \
0  360000  3032     4      4       2  1972    22221   29            0   
1  340000  2058     4      2       2  1976    22912   25            0   
2  250000  1780     4      3       2  1980    21345   21            0   
3  205500  1638     4      2       2  1963    17342   38            0   
4  275500  2196     4      3       2  1968    21786   33            0   

   quality_MEDIUM  style_2  style_3  style_4  style_5  style_6  style_7  \
0               1        0        0        0        0        0        0   
1               1        0        0        0        0        0        0   
2               1        0        0        0        0        0        0   
3               1        0        0        0        0        0        0   
4               1        0        0        0        0        0        1   

   style_9  highway_YES  pool_YES  ac_YES  
0        0            0         0       1  
1        0            0         0       1  
2        0            0         0       1  
3        0            0         0       1  
4        0            0         0       1

In [56]:
#dummy
X = credit1[['area','beds','baths','garage','age','style_2','style_3','style_4','style_5','style_6','style_7','style_9',
         'lotsize','ac_YES','pool_YES','quality_LOW','quality_MEDIUM','highway_YES']]
y = credit1[['price']]
X = sm.add_constant(X)
model = sm.OLS(y, X.astype(float)).fit()

In [67]:
list3 = ['constant','area','beds','baths','garage','age','style_2','style_3','style_4','style_5','style_6','style_7',
         'style_9','lotsize','ac_YES','pool_YES','quality_LOW','quality_MEDIUM','highway']
list3_v = [1,
           3150,
           2,
           3,2,2001-1996,0,0,0,0,0,0,0,26250,1,1,0,0,0]

In [68]:
x = pd.DataFrame([list3_v],columns = list3)
x

constant  area  beds  baths  garage  age  style_2  style_3  style_4  \
0         1  3150     2      3       2    5        0        0        0   

   style_5  style_6  style_7  style_9  lotsize  ac_YES  pool_YES  quality_LOW  \
0        0        0        0        0    26250       1         1            0   

   quality_MEDIUM  highway  
0               0        0

In [65]:
pred = model.get_prediction(x)
pred.summary_frame()

mean       mean_se  mean_ci_lower  mean_ci_upper  obs_ci_lower  \
0  585922.526796  15323.494616  555803.458271   616041.59532  479849.69376   

    obs_ci_upper  
0  691995.359831

#### Ans: 479849.69376 ~ 691995.359831